category_theory/limits/limits.lean

```diff
@@ -0,0 +1,584 @@
+is_colimit.mk_cocone_morphism
+  (λ s, i.inv ≫ P.desc_cocone_morphism s)
+  (λ s m, by rw i.eq_inv_comp; apply P.uniq_cocone_morphism)
+
+variables {t : cocone F}
+
+lemma is_colimit.hom_desc (h : is_colimit t) {W : C} (m : t.X ⟶ W) :
+  m = h.desc { X := W, ι := { app := λ b, t.ι.app b ≫ m,
+    naturality' := by intros; erw [←assoc, t.ι.naturality, comp_id, comp_id] } } :=
+h.uniq { X := W, ι := { app := λ b, t.ι.app b ≫ m, naturality' := _ } } m (λ b, rfl)
+
+/-- Two morphisms out of a colimit are equal if their compositions with
+  each cocone morphism are equal. -/
+lemma is_colimit.hom_ext (h : is_colimit t) {W : C} {f f' : t.X ⟶ W}
+  (w : ∀ j, t.ι.app j ≫ f = t.ι.app j ≫ f') : f = f' :=
+by rw [h.hom_desc f, h.hom_desc f']; congr; exact funext w
+
+/-- The universal property of a colimit cocone: a map `X ⟶ W` is the same as
+  a cocone on `F` with vertex `W`. -/
+def is_colimit.hom_iso (h : is_colimit t) (W : C) : (t.X ⟶ W) ≅ (F ⟹ (const J).obj W) :=
+{ hom := λ f, (t.extend f).ι,
+  inv := λ ι, h.desc { X := W, ι := ι },
+  hom_inv_id' := by ext f; apply h.hom_ext; intro j; simp; dsimp; refl }
+
+@[simp] lemma is_colimit.hom_iso_hom (h : is_colimit t) {W : C} (f : t.X ⟶ W) :
+  (is_colimit.hom_iso h W).hom f = (t.extend f).ι := rfl
+
+/-- The colimit of `F` represents the functor taking `W` to
+  the set of cocones on `F` with vertex `W`. -/
+def is_colimit.nat_iso (h : is_colimit t) : coyoneda.obj t.X ≅ F.cocones :=
+nat_iso.of_components (is_colimit.hom_iso h) (by intros; ext; dsimp; rw ←assoc; refl)
+
+def is_colimit.hom_iso' (h : is_colimit t) (W : C) :
+  (t.X ⟶ W) ≅ { p : Π j, F.obj j ⟶ W // ∀ {j j' : J} (f : j ⟶ j'), F.map f ≫ p j' = p j } :=
+h.hom_iso W ≪≫
+{ hom := λ ι,
+  ⟨λ j, ι.app j, λ j j' f,
+   by convert ←(ι.naturality f); apply comp_id⟩,
+  inv := λ p,
+  { app := λ j, p.1 j,
+    naturality' := λ j j' f, begin dsimp, erw [comp_id], exact (p.2 f) end } }
+
+end is_colimit
+
+
+section limit
+
+/-- `has_limit F` represents a particular chosen limit of the diagram `F`. -/
+class has_limit (F : J ⥤ C) :=
+(cone : cone F)
+(is_limit : is_limit cone)
+
+variables (J C)
+
+/-- `C` has limits of shape `J` if we have chosen a particular limit of
+  every functor `F : J ⥤ C`. -/
+@[class] def has_limits_of_shape := Π F : J ⥤ C, has_limit F
+
+/-- `C` has all (small) limits if it has limits of every shape. -/
+@[class] def has_limits :=
+Π {J : Type v} {𝒥 : small_category J}, by exactI has_limits_of_shape J C
+
+variables {J C}
+
+instance has_limit_of_has_limits_of_shape
+  {J : Type v} [small_category J] [H : has_limits_of_shape J C] (F : J ⥤ C) : has_limit F :=
+H F
+
+instance has_limits_of_shape_of_has_limits
+  {J : Type v} [small_category J] [H : has_limits.{u v} C] : has_limits_of_shape J C :=
+H
+
+/- Interface to the `has_limit` class. -/
+
+def limit.cone (F : J ⥤ C) [has_limit F] : cone F := has_limit.cone F
+
+def limit (F : J ⥤ C) [has_limit F] := (limit.cone F).X
+
```

Comment:
The following could all go into a `limit` namespace...

Reply:
I'm less sure about this one. I tried, but it feels like it's saving some characters at the expensive of some obscurity. I'm happy to have a second opinion.

Reply:
@rwbarton, do you think we should namespace all these? I'm on the fence. (But can do it if it's warranted.)